libcxx/include/__memory/indirect.h

```diff
@@ -0,0 +1,328 @@
+      return *this;
+
+    static constexpr bool __propagate_allocator =
+        allocator_traits<_Allocator>::propagate_on_container_copy_assignment::value;
+    if (__other.valueless_after_move()) {
+      __destroy_owned_object();
+      __ptr_ = nullptr;
+    } else if (!valueless_after_move() && __alloc_ == __other.__alloc_) {
+      *__ptr_ = *__other;
+    } else {
+      pointer __new_ptr;
+      if constexpr (__propagate_allocator) {
+        // We need a mutable instance of the allocator, so make a copy.
+        _Allocator __alloc_copy = __other.__alloc_;
+        __new_ptr               = __allocate_owned_object(__alloc_copy, *__other);
+      } else {
+        __new_ptr = __allocate_owned_object(__alloc_, *__other);
+      }
+      __destroy_owned_object();
+      __ptr_ = __new_ptr;
+    }
+
+    if constexpr (__propagate_allocator)
+      __alloc_ = __other.__alloc_;
+
+    return *this;
+  }
+
+  _LIBCPP_HIDE_FROM_ABI constexpr indirect& operator=(indirect&& __other) noexcept(
+      allocator_traits<_Allocator>::propagate_on_container_move_assignment::value ||
+      allocator_traits<_Allocator>::is_always_equal::value) {
+    if (std::addressof(__other) == this)
+      return *this;
+
+    static constexpr bool __propagate_allocator =
+        allocator_traits<_Allocator>::propagate_on_container_move_assignment::value;
+
+    pointer __new_ptr;
+    if constexpr (__propagate_allocator || allocator_traits<_Allocator>::is_always_equal::value) {
+      __new_ptr = __other.__ptr_;
+    } else if (__other.valueless_after_move()) {
+      __new_ptr = nullptr;
+    } else if (__alloc_ == __other.__alloc_) {
+      __new_ptr = __other.__ptr_;
+    } else {
+      __new_ptr = __allocate_owned_object(__alloc_, *std::move(__other));
+      __other.__destroy_owned_object();
+    }
+    __other.__ptr_ = nullptr;
+    __destroy_owned_object();
+    __ptr_ = __new_ptr;
+
+    if constexpr (__propagate_allocator)
+      __alloc_ = __other.__alloc_;
+
+    return *this;
+  }
+
+  template <class _Up = _Tp>
+    requires(!is_same_v<remove_cvref_t<_Up>, indirect> && is_constructible_v<_Tp, _Up> && is_assignable_v<_Tp&, _Up>)
+  _LIBCPP_HIDE_FROM_ABI constexpr indirect& operator=(_Up&& __u) {
+    if (valueless_after_move())
+      __ptr_ = __allocate_owned_object(__alloc_, std::forward<_Up>(__u));
+    else
+      *__ptr_ = std::forward<_Up>(__u);
+    return *this;
+  }
+
+  // [indirect.obs], observers
+  template <class _Self>
+    requires(!is_volatile_v<_Self>)
+  [[nodiscard]] _LIBCPP_HIDE_FROM_ABI constexpr auto&& operator*(this _Self&& __self) noexcept {
+    _LIBCPP_ASSERT_VALID_ELEMENT_ACCESS((!std::__forward_as<_Self, indirect>(__self).valueless_after_move()),
+                                        "operator* called on a valueless std::indirect object");
+    return std::forward_like<_Self>(*__self.__ptr_);
+  }
+
+  [[nodiscard]] _LIBCPP_HIDE_FROM_ABI constexpr const_pointer operator->() const noexcept {
```

Comment:
Does it make sense to mark operator-> as `[[nodiscard]]` please see: https://godbolt.org/z/andnYfWqn